Wrong console message's repeat count is updated when messages are mixed from page and Web Workers.

basuke opened the pull request:
#### 1a9813d49b1b047214403cd0b6a26ce46eabb270
<pre>
Wrong console message&apos;s repeat count is updated when messages are mixed from page and Web Workers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300728">https://bugs.webkit.org/show_bug.cgi?id=300728</a>
<a href="https://rdar.apple.com/162612099">rdar://162612099</a>

Reviewed by Devin Rousso.

Console messages are compared in each target in Web Content process. If the message is
same with previous message, it sends `messageRepeatCountUpdated` message to the frontend.
In frontend, it updates simply previously displayed messageView&apos;s repeat count.

If a Worker sends a message, the things get messy. Because comparison happens in each
target, they don&apos;t know each other. Only checks if the message is same with the one from
the target. But that message can be not the one in the frontend.

1. main frame send a message A.
2. Worker send a message B.
3. Now console UI have line A and B.
4. In main frame, send identical message A.
5. Before sending the message, ConsoleAgent compares the one with previously sent message at step 1.
6. Because they are identical, it send `messageRepeatCountUpdated` with count 2.
7. Frontend get `messageRepeatCountUpdated` and it tries to update previous message view, but that is view for B.
8. B has updated with [2] count label.

To fix this, keep storing the messageView for each target. When `updatePreviousMessageRepeatCount`
is called, it checks _previousMessageView has same target and if not, it creates the new
messageView with previous message to avoid updating the wrong message count.

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager.prototype.messageRepeatCountUpdated):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype.appendConsoleMessage):
(WI.JavaScriptLogViewController.prototype.updatePreviousMessageRepeatCount):
(WI.JavaScriptLogViewController.prototype._appendConsoleMessageView):
(WI.JavaScriptLogViewController.prototype._didRenderConsoleMessageView):
* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messageRepeatCountUpdated):
* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._previousMessageRepeatCountUpdated):

Canonical link: <a href="https://commits.webkit.org/301917@main">https://commits.webkit.org/301917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5448d9a5fc84e956ef3aa9e4a8c6fa24bc5ad017

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134436 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78927 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e1ec7c16-88da-441b-b556-70f10237bd91) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96937 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6ddc881d-29a3-4d7b-9462-bae7dec7561b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38120 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114083 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77431 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3288686d-d513-49d2-9cbe-52a6a7573cc3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36965 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32197 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77818 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32602 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136919 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54031 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105457 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54542 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110432 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105134 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26824 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50658 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29108 "Found 1 new test failure: http/tests/misc/ftp-eplf-directory.py (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51610 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53968 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60055 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53202 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56659 "Built successfully") | | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54961 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -157,8 +157,24 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            }
+        }
+        return null;
+    }
```
In connection to the suggestion to remove `this._consoleMessageViews` from https://github.com/WebKit/WebKit/pull/52404#discussion_r2433609094, what if we introduce a `WeakMap` to hold the last console message view per target?

Something like this (not tested):

```diff
diff --git a/Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js b/Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js
index 18707a81f33e..d0a28384b28e 100644
--- a/Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js
+++ b/Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js
@@ -69,6 +69,7 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
         this._consoleMessageViews = [];
         this._showTimestamps = WI.settings.showConsoleMessageTimestamps.value;
+        this._lastConsoleMessageForTargetMap = new WeakMap();
 
         this.startNewSession();
     }
@@ -111,6 +112,7 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
         let consoleSession = new WI.ConsoleSession(data); 
+        this._lastConsoleMessageForTargetMap = new WeakMap();
         this._previousMessageView = null;
         this._lastCommitted = {text: "", special: false};
         this._repeatCountWasInterrupted = false;
@@ -157,8 +159,10 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
         return consoleMessageView;
     }
 
-    updatePreviousMessageRepeatCount(count, timestamp)
+    updatePreviousMessageRepeatCount(target, count, timestamp)
     {
+        this._previousMessageView = this._lastConsoleMessageForTargetMap.get(target);
+
         console.assert(this._previousMessageView);
         if (!this._previousMessageView)
             return false;
@@ -289,6 +293,7 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
         }
         pendingMessagesForSession.push(messageView);
         this._consoleMessageViews.push(messageView);
+        this._lastConsoleMessageForTargetMap.set(messageView.message.target, messageView);
 
         this._cleared = false;
         this._repeatCountWasInterrupted = repeatCountWasInterrupted || false;

```

We could probably drop `this._previousMessageView` completely. 

Would this solve the need?
[consoleMessageViewForTarget.patch](https://github.com/user-attachments/files/22935520/consoleMessageViewForTarget.patch)

basuke replied:
That's smart solution!

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -157,8 +157,24 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+    findPreviousMessageViewForTarget(target)
     {
+        var index = this._consoleMessageViews.length;
+        while (--index >= 0) {
```
```suggestion
        for (let index = this._consoleMessageViews.length - 1; index >= 0; --index) {
```

the-chenergy replied:
Or even rename `index` to `i`, given its short lifespan.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -44,7 +44,7 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
         this._cleared = true;
-        this._previousMessageView = null;
+        this._lastConseoleMessageViewForTargetMap = new WeakMap();
```
typo in the name "Conseole" (and also there's no need for the "Map" since the "fooForBar" pattern implies that it's a `Map`)

Style: we dont normally add `()` when there's no arguments with `new`

```js
this._lastConsoleMessageViewForTarget = new WeakMap;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -111,7 +111,7 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
         let consoleSession = new WI.ConsoleSession(data);
 
-        this._previousMessageView = null;
+        this._lastConseoleMessageViewForTargetMap = new WeakMap();
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -157,22 +157,24 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
     {
-        console.assert(this._previousMessageView);
-        if (!this._previousMessageView)
+        const previousMessageView = this._lastConseoleMessageViewForTargetMap.get(target);
```
Style: we only use `const` for things that are actually truly constant (e.g. `const foo = 42;`) so please use `let` instead

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -157,22 +157,24 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
         }
 
-        var consoleMessage = this._previousMessageView.message;
+        const consoleMessage = previousMessageView.message;
```
ditto to use `let`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -293,16 +295,18 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
-        if (!repeatCountWasInterrupted)
-            this._previousMessageView = messageView;
+        const message = messageView.message;
```
ditto to use `let`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -293,16 +295,18 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
-            this._previousMessageView = messageView;
+        const message = messageView.message;
+
+        if (!repeatCountWasInterrupted && message && message.target)
```
it should not be possible to create a `WI.ConsoleMessage` without a `WI.Target` or a `WI.ConsoleMessageView` without a `WI.ConsoleMessage` so these extra checks should not be necessary
```js
if (!repeatCountWasInterrupted)
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -293,16 +295,18 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
             this.renderPendingMessagesSoon();
 
-        if (!WI.isShowingConsoleTab() && messageView.message && messageView.message.shouldRevealConsole)
+        if (!WI.isShowingConsoleTab() && message && message.shouldRevealConsole)
```
ditto about `message` always existing
```js
if (!WI.isShowingConsoleTab() && message.shouldRevealConsole)
```

rcaliman-apple replied:
> ditto about `message` always existing
> 
> ```js
> if (!WI.isShowingConsoleTab() && message.shouldRevealConsole)
> ```

@dcrousso, @basuke  Turns out the checks were intentional. The messageView could also be a `WI.ConsoleCommandView` (instead of `WI.ConsoleMessageView`). Fix in https://github.com/WebKit/WebKit/pull/52873

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -520,7 +520,7 @@ WI.LogContentView = class LogContentView extends WI.ContentView
     _previousMessageRepeatCountUpdated(event)
     {
-        if (!this._logViewController.updatePreviousMessageRepeatCount(event.data.count, event.data.timestamp))
+        if (!this._logViewController.updatePreviousMessageRepeatCount(event.data.target, event.data.count, event.data.timestamp))
```
NIT: rather than repeating `event.data` i think we should extract it earlier
```js
let {target, count, timestamp} = event.data;
if (!this._logViewController.updatePreviousMessageRepeatCount(target, count, timestamp))
    return;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -293,16 +295,18 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
-        if (messageView.message && messageView.message.source !== WI.ConsoleMessage.MessageSource.JS)
+        if (message && message.source !== WI.ConsoleMessage.MessageSource.JS)
             this._lastCommitted = {test: "", special: false};
```
Aside: this should actually be `{text: "", special: false}` if you want to fix it as a drive-by :)

dcrousso reviewed (changes requested):
while all of this does look good (with a few Style and NIT) i think it's the wrong approach

i dont think we want to preserve the last message for each target as the point of the repeat count is just to avoid repeating the overall last message in the entire console

for example, it would be weird if some message that's way earlier in the console suddenly started incrementing it's repeat count instead of being newly logged at the end of the view

put another way, the fact that the message comes from entire separate WebProcess is an implementation detail so the developer shouldn't care about or be aware of that at all

as such, i think we keep this logic to just remember the last message (regardless of the target) and have the frontend generate a visual copy of the `WI.ConsoleMessage` with a new repeat count of `0` (which i think `WI.JavaScriptLogViewController` already does)

you should be able to use/adjust/etc. the existing logic for handling multiple `Worker` logging at the same time (e.g. always set `this._repeatCountWasInterrupted = true` whenever a new message is added even if it's from the same `target`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -154,22 +156,32 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
-        var previousIgnoredCount = this._previousMessageView[WI.JavaScriptLogViewController.IgnoredRepeatCount] || 0;
-        var previousVisibleCount = this._previousMessageView.repeatCount;
+        if (previousMessageView.message.target._identifier !== target._identifier) {
```
i think we can compare `previousMessageView.message.target !== target` directly since we should never have more than one `WI.Target` instance for a given `identifier`

at the very least we should be using `target.identifier` public `get` instead of the private/internal property

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -154,22 +156,32 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            console.assert(previousMessageView);
+            if (!previousMessageView)
+                return false;
+            this.appendConsoleMessage(previousMessageView.message);
```
should we be setting `this._repeatCountWasInterrupted = true;` before adding the new message since the repeat count was actually interrupted (by a message from another target)?

basuke replied:
Right! Now I fully understood the purpose of this property.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -154,22 +156,32 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+        }
+
+        var previousIgnoredCount = previousMessageView[WI.JavaScriptLogViewController.IgnoredRepeatCount] || 0;
+        var previousVisibleCount = previousMessageView.repeatCount;
```
Style: we can update these to `let` since we're changing these lines anyways

dcrousso reviewed (approved):
nice fix!  thanks for iterating on the solution with me :)

basuke committed:
Wrong console message's repeat count is updated when messages are mixed from page and Web Workers.
https://bugs.webkit.org/show_bug.cgi?id=300728
rdar://162612099

Reviewed by Devin Rousso.

Console messages are compared in each target in Web Content process. If the message is
same with previous message, it sends `messageRepeatCountUpdated` message to the frontend.
In frontend, it updates simply previously displayed messageView's repeat count.

If a Worker sends a message, the things get messy. Because comparison happens in each
target, they don't know each other. Only checks if the message is same with the one from
the target. But that message can be not the one in the frontend.

1. main frame send a message A.
2. Worker send a message B.
3. Now console UI have line A and B.
4. In main frame, send identical message A.
5. Before sending the message, ConsoleAgent compares the one with previously sent message at step 1.
6. Because they are identical, it send `messageRepeatCountUpdated` with count 2.
7. Frontend get `messageRepeatCountUpdated` and it tries to update previous message view, but that is view for B.
8. B has updated with [2] count label.

To fix this, keep storing the messageView for each target. When `updatePreviousMessageRepeatCount`
is called, it checks _previousMessageView has same target and if not, it creates the new
messageView with previous message to avoid updating the wrong message count.

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager.prototype.messageRepeatCountUpdated):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype.appendConsoleMessage):
(WI.JavaScriptLogViewController.prototype.updatePreviousMessageRepeatCount):
(WI.JavaScriptLogViewController.prototype._appendConsoleMessageView):
(WI.JavaScriptLogViewController.prototype._didRenderConsoleMessageView):
* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messageRepeatCountUpdated):
* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._previousMessageRepeatCountUpdated):

Canonical link: https://commits.webkit.org/301917@main
